credentials/alts/internal/handshaker/handshaker.go

```diff
@@ -198,6 +194,14 @@ func (h *altsHandshaker) ClientHandshake(ctx context.Context) (net.Conn, credent
 		return nil, nil, errors.New("only handshakers created using NewClientHandshaker can perform a client handshaker")
 	}
 
+	if h.stream == nil {
```

Comment:
Is not an invariant that `h.stream == nil` at this point?

If so should we change this conditional to an assertion that `stream` is nil ?

Reply:
Unfortunately we can't do that right now, because of how the unit tests in handshaker_test.go are set up: they inject a fake stream (using non-exported APIs) and, if we unconditionally set `h.stream` based on `h.clientConn`, then the tests will fail (since we cannot modify `altsgrpc.NewHandshakerServiceClient` to return the fake stream AFAIK).

The solution to this is to overhaul the unit tests to only use the exported APIs and talk to a fake handshaker service over a real grpc channel. This will be a large-ish change and (IMO) should be out-of-scope for this PR, which should just focus on the bug fix. I've added TODOs to come back and overhaul the unit tests in a follow-up PR.

WDYT? Does that sound reasonable?

Reply:
SGTM